cli/spcli/sectors.go

```diff
@@ -876,84 +877,97 @@ func SectorsExtendCmd(getActorAddress ActorAddressGetter) *cli.Command {
-										fmt.Printf("skipping sector %d because claim %d (client f0%s, piece %s) does not live long enough \n", sectorNumber, claimId, claim.Client, claim.Data)
-										cannotExtendSector = true
-										break
+									claimExpiration := claim.TermStart + claim.TermMax
+									// can be maintained in the extended sector
+									if claimExpiration > newExp {
+										claimIdsToMaintain = append(claimIdsToMaintain, claimId)
+									} else {
+										sectorInfo, ok := activeSectorsInfo[sectorNumber]
+										if !ok {
+											return xerrors.Errorf("failed to find sector in active sector set: %w", err)
+										}
+										if !cctx.Bool("drop-claims") ||
+											// FIP-0045 requires the claim minimum duration to have passed
+											currEpoch <= (claim.TermStart+claim.TermMin) ||
+											// FIP-0045 requires the sector to be in its last 30 days of life
+											(currEpoch <= sectorInfo.Expiration-builtin.EndOfLifeClaimDropPeriod) {
+											fmt.Printf("skipping sector %d because claim %d (client f0%s, piece %s) does not live long enough \n", sectorNumber, claimId, claim.Client, claim.Data)
+											cannotExtendSector = true
+											break
+										}
+
+										claimIdsToDrop = append(claimIdsToDrop, claimId)
 									}
 
-									claimIdsToDrop = append(claimIdsToDrop, claimId)
+									numbersToExtend = append(numbersToExtend, sectorNumber)
+								}
+								if cannotExtendSector {
+									continue
 								}
 
-								numbersToExtend = append(numbersToExtend, sectorNumber)
-							}
-							if cannotExtendSector {
-								continue
+								if len(claimIdsToMaintain)+len(claimIdsToDrop) != 0 {
+									sectorsWithClaims = append(sectorsWithClaims, miner.SectorClaim{
+										SectorNumber:   sectorNumber,
+										MaintainClaims: claimIdsToMaintain,
+										DropClaims:     claimIdsToDrop,
+									})
+								}
 							}
+						}
 
-							if len(claimIdsToMaintain)+len(claimIdsToDrop) != 0 {
-								sectorsWithClaims = append(sectorsWithClaims, miner.SectorClaim{
-									SectorNumber:   sectorNumber,
-									MaintainClaims: claimIdsToMaintain,
-									DropClaims:     claimIdsToDrop,
-								})
-							}
+						sectorsWithoutClaimsCount, err := sectorsWithoutClaimsToExtend.Count()
+						if err != nil {
+							return xerrors.Errorf("failed to count cc sectors: %w", err)
 						}
-					}
 
-					sectorsWithoutClaimsCount, err := sectorsWithoutClaimsToExtend.Count()
-					if err != nil {
-						return xerrors.Errorf("failed to count cc sectors: %w", err)
-					}
+						sectorsInDecl := int(sectorsWithoutClaimsCount) + len(sectorsWithClaims)
+						scount += sectorsInDecl
 
-					sectorsInDecl := int(sectorsWithoutClaimsCount) + len(sectorsWithClaims)
-					scount += sectorsInDecl
+						if len(p.Extensions) >= declMax {
+							params = append(params, p)
+							p = miner.ExtendSectorExpiration2Params{}
+							scount = sectorsInDecl
+						}
```

Comment:
I would suggest a complete refactor to get rid of this. If we are batching then we simply should determine the batch size correctly and apply the policies afterward. 

```suggestion
						if len(p.Extensions) >= declMax {
							params = append(params, p)
							p = miner.ExtendSectorExpiration2Params{}
							scount = sectorsInDecl
						}
```

Reply:
I can only help him get back on the right path, but I can't reconstruct it.